integration_test/third_party_apps_data/applications/dcgm/centos_rhel/install

```diff
@@ -1,54 +1,101 @@
-sudo bash ./NVIDIA-Linux-x86_64-$DRIVER_VERSION.run --silent
+verify_driver() {
+    # Verify NVIDIA driver:
+    # Installation could finish successfully but the driver is still unusable
+    # A common error when running this check:
+    # "NVIDIA-SMI has failed because it couldn't communicate with the NVIDIA 
+    # driver. Make sure that the latest NVIDIA driver is installed and running."
+    nvidia-smi
+}
 
-# check NVIDIA driver installation succeeded
-nvidia-smi
+install_driver_from_runfile() {
+    # Ref: https://docs.nvidia.com/datacenter/tesla/tesla-installation-notes/index.html#runfile
+    # This method requires the matching kernel-devel package to be installed, and 
+    # the package may be absent from the repo and cause this method to fail
+    # Remove existing installation before using the runfile
+    remove_driver_package
+    sudo yum install -y kernel-devel-$(uname -r) pciutils gcc make wget yum-utils 
+    local DRIVER_VERSION=535.129.03
+    echo "Installing NVIDIA Data Center driver $DRIVER_VERSION"
+    curl -fSsl -O https://us.download.nvidia.com/tesla/$DRIVER_VERSION/NVIDIA-Linux-x86_64-$DRIVER_VERSION.run
+    sudo bash ./NVIDIA-Linux-x86_64-$DRIVER_VERSION.run --silent
+    verify_driver
+}
 
-# Install DCGM
-VERSION_ID=${VERSION_ID%%.*}
-sudo yum-config-manager \
-    --add-repo http://developer.download.nvidia.com/compute/cuda/repos/rhel$VERSION_ID/x86_64/cuda-rhel$VERSION_ID.repo
-sudo yum clean expire-cache
-sudo yum install -y datacenter-gpu-manager
-sudo systemctl --now enable nvidia-dcgm
-
-# check DCGM service running and load profiling module
-dcgmi discovery --list
+setup_repo() {
+    sudo yum install -y yum-utils
+    sudo yum-config-manager \
+        --add-repo https://developer.download.nvidia.com/compute/cuda/repos/rhel$VERSION_ID/x86_64/cuda-rhel$VERSION_ID.repo
+    sudo yum clean all
+}
+
+install_driver_from_package_manager() {
+    setup_repo
+    install_driver_package
+    verify_driver
+}
+
+install_dcgm() {
+    # Ref: https://docs.nvidia.com/datacenter/dcgm/latest/user-guide/getting-started.html#rhel-centos-rocky-linux
+    setup_repo
+    sudo yum install -y datacenter-gpu-manager
+    sudo systemctl --now enable nvidia-dcgm 
+
+    # check DCGM service running and load profiling module
+    dcgmi discovery --list
+}
+
+try_install() {
+    for install_method in "$@"; do
+        echo "Installing NVIDIA driver with $install_method..."
+        # Temporarily disable the "-e" option so that failures in the function
+        # won't cause this script to exit; also run the installation method in a
+        # subshell with -e so that the function itself will exit immediately if 
+        # any command encounters error
+        set +e ; ( set -e; $install_method ) ; exit_code=$?
+        set -e
+        if [ $exit_code -eq 0 ]; then
+            echo "NVIDIA driver has been installed successfully with $install_method."
+            return 0
+        fi
```

Comment:
[Optional] Revisiting this, see the other thread (which I, sadly, can't make part of this review):
```suggestion
        # Can't use a subshell because of https://lists.gnu.org/archive/html/bug-bash/2012-12/msg00094.html
        bash -$- -c $install_method && {
            echo "NVIDIA driver has been installed successfully with $install_method."
            return 0
        }
```
[Some supporting experimentation](https://tio.run/##vVFdasMwDH7PKYQppWH4Ag07TGrLscCVh@2wjrGzp3LXtSVLoYOyJ9vS9yN93vXZT5Mb2RSKDK6nQFw87snagJsWPhsAND6CAijpg3jowIV@yK8rraTn@pDxBkMDx4T2pDQmVM1Xg4e3mApoN5dvNhmljt2ZzRHW663q5rifviULHAvggYpq5@x3Kn6ZL9UzJo/GIFq06g@a5Komud9jyZUXlB09pi7BBDIk7ZQqotpUxIsgLtA7WVz92m@SfjCmnfx4Naon6JUGbZ6XVxwLJqHL@z92uw7qiCl7Ydyf6KmLLzhP0xE "Bash – Try It Online").

You would also need to export all necessary functions at the top level for the `bash` subprocess to inherit them:
```
export -f install_driver_from_package_manager install_driver_from_runfile setup_repo install_driver_package remove_driver_package verify_driver 
```

Reply:
Done!

Used the second command to export all functions; also exported the `VERSION_ID` env. Thanks!